src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php

```
@@ -49,235 +52,296 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
     public function handle(AddCartRuleToOrderCommand $command): void
     {
         $order = $this->getOrderObject($command->getOrderId());
+        $orderInvoice = $this->createSingleOrderInvoice($order, $command);
 
-        // If the discount is for only one invoice
-        if ($order->hasInvoice() && null !== $command->getOrderInvoiceId()) {
-            $orderInvoice = new OrderInvoice($command->getOrderInvoiceId()->getValue());
-            if (!Validate::isLoadedObject($orderInvoice)) {
-                throw new OrderException('Can\'t load Order Invoice object');
+        if (null !== $orderInvoice) {
+            $orderInvoices = [$orderInvoice];
+        } else {
+            $orderInvoices = $order->getInvoicesCollection()->getResults();
+        }
+
+        $discountValue = (float) $command->getDiscountValue()->toPrecision(CommonAbstractType::PRESTASHOP_DECIMALS);
+        $reductionValues = $this->getReductionValues($command->getCartRuleType(), $discountValue, $order);
+        $reductionValuesByInvoiceId = $this->getReductionValuesByInvoiceId(
+            $reductionValues,
+            $command->getCartRuleType(),
+            $discountValue,
+            $orderInvoices
+        );
+        $this->updateInvoicesDiscount($orderInvoices, $reductionValuesByInvoiceId);
+
+        $result = true;
+
+        foreach ($reductionValuesByInvoiceId as &$reducedValues) {
+            $cartRuleObj = $this->createCartRule($command, $order, $reducedValues, $discountValue);
+
+            if ($result = $cartRuleObj->add()) {
+                $reducedValues['cart_rule_id'] = $cartRuleObj->id;
+            } else {
+                break;
```

This `break` happens when a cart rule could not be added, then `$result == false` and an exception is thrown L95
To make the code clearer could you please throw the exception here instead of the `break` ?